levels/level_manager.go

```diff
@@ -1707,6 +1727,84 @@ func containsTable(seg *segment, tabID sst.SSTableID) bool {
 	return false
 }
 
+func getSegmentForRegistration(segmentEntries []segmentEntry, registration RegistrationEntry) int {
+	n := len(segmentEntries)
+	// if n == 0 then sort.Search considers 0 a legal insertion point
+	// for now we still return -1 to fit in with the current code
+	if n == 0 {
+		return -1
+	}
+	// fits the case i == 0 && i == len(segmentEntries)-1
+	if n == 1 {
+		return 0
+	}
+	// with sort.Search we need a compare function that partitions the array into true/false
+	index := sort.Search(n, func(i int) bool {
+		return i == n-1 || bytes.Compare(registration.KeyEnd, segmentEntries[i+1].rangeStart) < 0
+	})
+	// now we can check if the KeyStart is > than segmentEntries[i-1].rangeEnd
+	if index == 0 {
+		return index
+	}
+	if bytes.Compare(registration.KeyStart, segmentEntries[index-1].rangeEnd) > 0 {
+		if index == n {
+			return index - 1
+		} else {
+			return index
+		}
+	}
+	return -1
+}
+
+func getSegmentEntryForDeregistration(segmentEntries []segmentEntry, deRegistration RegistrationEntry) int {
+	n := len(segmentEntries)
+	if n == 0 {
+		return -1
+	}
+	index := sort.Search(n, func(i int) bool {
+		return bytes.Compare(deRegistration.KeyStart, segmentEntries[i].rangeStart) < 0 ||
+			(bytes.Compare(deRegistration.KeyStart, segmentEntries[i].rangeStart) >= 0 &&
+				bytes.Compare(deRegistration.KeyEnd, segmentEntries[i].rangeEnd) <= 0)
+	})
+	// if index < n it's possible that sort.Search will consider it a legal index
+	// but we still need to test the equality condition
+	if index < n && bytes.Compare(deRegistration.KeyStart, segmentEntries[index].rangeStart) >= 0 &&
+		bytes.Compare(deRegistration.KeyEnd, segmentEntries[index].rangeEnd) <= 0 {
+		return index
+	}
+	return -1
+}
+
+func getTableEntryForDeregistration(seg *segment, deRegistration RegistrationEntry) int {
+	pos := -1
+	if deRegistration.Level == 0 {
+		for i, te := range seg.tableEntries {
+			if bytes.Equal(te.SSTableID, deRegistration.TableID) {
+				pos = i
+				break
+			}
+		}
+		return pos
+	}
+	n := len(seg.tableEntries)
+	// in principle if the ids are equal their ranges should be as well
+	pos = sort.Search(n, func(i int) bool {
+		return bytes.Compare(seg.tableEntries[i].RangeStart, deRegistration.KeyStart) >= 0
```

Comment:
Shouldn't it be == 0 ?
As we know the correct entry will have exactly matching range

Reply:
Because `Search` only finds the insertion point, the compare functio needs to partition the boundaries of the search into true/false. This is the implementation of `Search`

```go
```

Reply:
Sorry if I am being slow.
According to the documentation for search:
```
Search uses binary search to find and return the smallest index i
// in [0, n) at which f(i) is true
```
If we use a function that checks the range exactly using == 0, then it will return the index of the first item where that function is true, which will be the entry we want. What am I missing here?

Reply:
Here's a playground link that shows the difference https://go.dev/play/p/J9CZyYdU04q

the first one uses `>= 1` and we see the `h` values going 5 -> 2 > 1 -> 0, in the second one using `== 1`, we see the `h` values going 5 -> 8 -> 9.

I don't think the documentation is wrong, just not very clear